### PR TITLE
feat(all): #1096 ssr

### LINKED ISSUE #1096: Evaluate Server Side Rendering (SSR) capabilities of lit with example applicaiton
### Description / User story

We want to evaluate the server side rendering (SSR) capabilities of lit via an example application, before we look deeper into CSS improvements as defined via #1081 .

The goal of this task is to have a better understanding of the SSR requirements, which also effect the decision for our CSS approach.

### Requirements / Prerequisites

- none

### Acceptance Criteria

- [x] We have a better understanding of the SSR requirements and are able to decide on the CSS approach after further research of Emotion, Vanilla Extract and SASS has been finished
- [ ] The outcome of this evaluation has been documented
- [x] The `js-example-app` is extended to demonstrate SSR capabilities and client side hydration, if possible

### Additional information

- ...

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by faselbaum**
# Changes

## `js-example-app` SSR
To demonstrate `lit`'s server side rendering capabilities the `js-example-app` package was converted to a NodeJs [Koa](https://github.com/koajs/koa) application. This includes client side hydration. The ability to perform a static build has been removed. As such the render.com deployment configuration needs to be updated. 

## Transition to ESM
Library packages such as `@boiler/ui-library` and `@boiler/icons` have been transitioned to "ESM only" to reduce complexity and enable a unified development approach on client and server.

To be ESM compliant all import specifiers have to be explicit. As a result all implicit imports have been adapted.

For now all bundle steps have been removed. Bundlers can be re-introduced should the need for minified production bundles arise. We now rely on `tsc` only. This yields accurate type definitions, source maps and makes debugging much easier since we don't ship garbled js to our consuming packages.

## `lit` upgrade to 3.x
The `@lit-labs/ssr` package requires version of `lit@^3.0.0`. So all our lit dependencies had to be upgraded.

Furthermore `lit@^3.0.0` dropped support for legacy decorators. This necessitates the switch to the new `accessor` syntax for decorated class fields. ([lit docs](https://lit.dev/docs/components/decorators/#migrating-typescript-standard-decorators)) 

> https://lit.dev/docs/components/decorators/#decorators-babel
> Note: Lit decorators only work with "version": "2023-05". Other versions, including the formerly supported "2018-09", are not supported.

The above is relevant for Storybook, which uses babel internally to transpile TS source files.

## `@boiler/icons`
The icon generation step for the `@boiler/icons` package has been changed to inline all SVG icons as strings into one JavaScript icon module instead of relying on webpack's [raw-loader](https://v4.webpack.js.org/loaders/raw-loader/#inline) syntax. This enables the use of icons from vanilla ES modules without the need for a bundler on consumer side.

Optimized icons have been removed from source control and are now being generated during the compile step.

---
Closes #1096

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,9 @@
 # B01LER
-![B01LER_UpdateImg_1100-480@2x](https://github.com/deven-org/boiler/assets/122102805/7f55e1fd-955f-4c9f-890d-3d5a03f5c3aa)
 
+![B01LER_UpdateImg_1100-480@2x](https://github.com/deven-org/boiler/assets/122102805/7f55e1fd-955f-4c9f-890d-3d5a03f5c3aa)
 
 ## :page_with_curl: Content
+
```
Comment (angsherpa456) [inline README.md:6]: .md file does not have the new line after the heading, please for consistency purporse, remove the newly added lines throughout the file. New line might have been added by the vscode, prettier-code formatter extension.

File: docs/CONTRIBUTING.md [hunk 2]
```diff
@@ -1,4 +1,5 @@
 # Contributing
+
```
Comment (angsherpa456) [inline docs/CONTRIBUTING.md:2]: Same as above comment.

File: packages/js-example-app/src/index.client.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  BlrInputFieldNumber,
+  BlrInputFieldText,
+  BlrRadio,
+  BlrSelect,
+  BlrTabBar,
+  BlrTextarea,
+  BlrToggleSwitch,
+} from '@boiler/ui-library';
+
+await hydrate();
```
Comment (angsherpa456) [inline packages/js-example-app/src/index.client.ts:14]: Functionality might be ok but simply for the properness of the code flow, declare the functions first and then function call later on. Same for line 15.

File: packages/js-example-app/src/index.client.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+    // Apply the polyfill. This is a one-shot operation, so it is important
+    // it happens after all HTML has been parsed.
+    hydrateShadowRoots(document.body);
+  }
+
+  await litHydrateSupportInstalled;
+
+  // Import component modules causing them to become interactive
+  await import('@boiler/ui-library');
```
Comment (angsherpa456) [inline packages/js-example-app/src/index.client.ts:34]: I have noticed that the component gets loaded visually but the functionality is not there yet ( clicking on 'Toggle loading state' does not do anything), might be that the hydration has not completed yet at this point even though the elements are already painted. In this case, is there a way to wait and show the loading text or something similar while the hydration is still going on.

File: packages/js-example-app/src/index.server.ts [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+                variant="primary"
+                size-variant="md"
+                icon="blr360"
+                aria-label="Button Icon"
+                button-icon-id="buttonIconId"
+              >
+              </blr-button-icon>
+            </div>
+
+            <div class="component">
```
Comment (angsherpa456) [inline packages/js-example-app/src/index.server.ts:52]: Currently not working. Need to explore the reason.

File: packages/js-example-app/src/index.server.ts [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+                value=""
+                label="Label"
+                aria-label=""
+                radio-id="radioId"
+                name="Radio Button"
+              >
+              </blr-radio>
+            </div>
+
+            <div class="component">
```
Comment (angsherpa456) [inline packages/js-example-app/src/index.server.ts:188]: Currently not working. Need to explore the reason.

File: packages/js-example-app/src/server.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import Koa from 'koa';
+import { AddressInfo } from 'node:net';
+import Router from '@koa/router';
+import staticFiles from 'koa-static';
+import { nodeResolve } from 'koa-node-resolve';
+import { RenderResultReadable } from '@lit-labs/ssr/lib/render-result-readable.js';
+
```
Comment (angsherpa456) [inline packages/js-example-app/src/server.ts:7]: Remove this line please

File: packages/ui-library/src/components/loader/index.ts [hunk 8]
```diff
@@ -1,19 +1,20 @@
...
+import { staticStyles } from './index.css.js';
 
-import { TAG_NAME } from './renderFunction';
-import { ThemeType } from '../../foundation/_tokens-generated/index.themes';
-import { ActionSizesType, FeedbackVariantType } from '../../globals/types';
-import { LitElementCustom } from '../../utils/lit-element-custom';
+import { TAG_NAME } from './renderFunction.js';
+import { ThemeType } from '../../foundation/_tokens-generated/index.themes.js';
+import { ActionSizesType, FeedbackVariantType } from '../../globals/types.js';
+import { LitElementCustom } from '../../utils/lit/element.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/loader/index.ts:9]: Please integrate the line 9 and 10

File: packages/ui-library/src/components/radio-group/index.ts [hunk 9]
```diff
@@ -1,43 +1,44 @@
...
-import { LitElementCustom } from '../../utils/lit-element-custom';
+import { ThemeType } from '../../foundation/_tokens-generated/index.themes.js';
+import { staticStyles as staticRadioStyles } from '../../foundation/component-tokens/radio.css.js';
+import { staticStyles as staticFormStyles } from '../../foundation/semantic-tokens/form.css.js';
+import { InputSizesType, RadioGroupDirection, RadioOption } from '../../globals/types.js';
+import { BlrFormCaptionGroupRenderFunction } from '../form-caption-group/renderFunction.js';
+import { BlrFormCaptionRenderFunction } from '../form-caption/renderFunction.js';
+import { BlrFormLabelInlineRenderFunction } from '../form-label/form-label-inline/renderFunction.js';
+import { TAG_NAME } from './renderFunction.js';
+import { LitElementCustom } from '../../utils/lit/element.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/radio-group/index.ts:14]: Please integrate line 14 and 15

File: packages/ui-library/src/components/radio/index.ts [hunk 10]
```diff
@@ -1,43 +1,45 @@
...
-import { BlrFormCaptionRenderFunction } from '../form-caption/renderFunction';
-import { BlrFormLabelInlineRenderFunction } from '../form-label/form-label-inline/renderFunction';
-import { createBlrBlurEvent, createBlrFocusEvent, createBlrSelectedValueChangeEvent } from '../../globals/events';
-import { LitElementCustom } from '../../utils/lit-element-custom';
+import { ThemeType } from '../../foundation/_tokens-generated/index.themes.js';
+import { BlrFormCaptionGroupRenderFunction } from '../form-caption-group/renderFunction.js';
+import { BlrFormCaptionRenderFunction } from '../form-caption/renderFunction.js';
+import { BlrFormLabelInlineRenderFunction } from '../form-label/form-label-inline/renderFunction.js';
+import { createBlrBlurEvent, createBlrFocusEvent, createBlrSelectedValueChangeEvent } from '../../globals/events.js';
+import { LitElementCustom } from '../../utils/lit/element.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/radio/index.ts:15]: Pleaes integrate line 15 and 16

File: packages/js-example-app/src/index.client.ts (hunk 4 above)
Comment (faselbaum) [reply packages/js-example-app/src/index.client.ts:31]: Noted in new [Issue](https://github.com/deven-org/boiler/issues/1122#issue-2348276750). See point **A)**

File: packages/ui-library/src/components/slider-single-value/range-legend-slider/index.ts [hunk 11]
```diff
@@ -1,45 +1,47 @@
...
-import { FormSizesType, ActionVariantType, RenderBtnProps } from '../../../globals/types';
-import { setOnclickValue, findToolTipPosition } from '../../../utils/range-slider-utils';
-import { BlrButtonIconRenderFunction } from '../../button-icon/renderFunction';
-import { LitElementCustom } from '../../../utils/lit-element-custom';
+import { ThemeType } from '../../../foundation/_tokens-generated/index.themes.js';
+import { staticStyles as staticSharedStyles } from '../../../foundation/component-tokens/slider-legend.css.js';
+import { FormSizesType, ActionVariantType, RenderBtnProps } from '../../../globals/types.js';
+import { setOnclickValue, findToolTipPosition } from '../../../utils/range-slider-utils.js';
+import { BlrButtonIconRenderFunction } from '../../button-icon/renderFunction.js';
+import { LitElementCustom } from '../../../utils/lit/element.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/slider-single-value/range-legend-slider/index.ts:16]: Please integrate line 16 and 17

File: packages/js-example-app/src/index.server.ts (hunk 5 above)
Comment (faselbaum) [reply packages/js-example-app/src/index.server.ts:52]: Noted in new [Issue](https://github.com/deven-org/boiler/issues/1122#issue-2348276750). See point **B)**

File: packages/js-example-app/src/index.server.ts (hunk 6 above)
Comment (faselbaum) [reply packages/js-example-app/src/index.server.ts:188]: Noted in new [Issue](https://github.com/deven-org/boiler/issues/1122#issue-2348276750). See point **B)**

File: packages/ui-library/src/components/slider-single-value/range-slider/index.ts [hunk 12]
```diff
@@ -1,45 +1,52 @@
...
-import { LitElementCustom } from '../../../utils/lit-element-custom';
+import { ThemeType } from '../../../foundation/_tokens-generated/index.themes.js';
+import {
+  findPercentage,
+  generateRangeBar,
+  setOnclickValue,
+  findNearestValue,
+} from '../../../utils/range-slider-utils.js';
+import { BlrButtonIconRenderFunction } from '../../button-icon/renderFunction.js';
+import { LitElementCustom } from '../../../utils/lit/element.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/slider-single-value/range-slider/index.ts:19]: Please integrate line 19 and 20

File: packages/ui-library/src/components/slider-single-value/range-slider/renderFunction.ts [hunk 13]
```diff
@@ -1,6 +1,6 @@
-// import { TAG_NAME } from '.';
-import type { BlrRangeSliderType } from '.';
-import { genericBlrComponentRenderer } from '../../../utils/typesafe-generic-component-renderer';
+// import { TAG_NAME } from './index.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/slider-single-value/range-slider/renderFunction.ts:1]: If not required please remove it

File: packages/ui-library/src/components/slider-two-values/range-legend-min-max-slider/index.ts [hunk 14]
```diff
@@ -1,46 +1,48 @@
...
-import { FormSizesType, ActionVariantType, RenderBtnProps } from '../../../globals/types';
-import { setOnclickValue, findToolTipPosition } from '../../../utils/range-slider-utils';
-import { BlrButtonIconRenderFunction } from '../../button-icon/renderFunction';
-import { LitElementCustom } from '../../../utils/lit-element-custom';
+import { ThemeType } from '../../../foundation/_tokens-generated/index.themes.js';
+import { staticStyles as staticSharedStyles } from '../../../foundation/component-tokens/slider-legend.css.js';
+import { FormSizesType, ActionVariantType, RenderBtnProps } from '../../../globals/types.js';
+import { setOnclickValue, findToolTipPosition } from '../../../utils/range-slider-utils.js';
+import { BlrButtonIconRenderFunction } from '../../button-icon/renderFunction.js';
+import { LitElementCustom } from '../../../utils/lit/element.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/slider-two-values/range-legend-min-max-slider/index.ts:15]: Please integrate line 15 and 16

File: packages/ui-library/src/components/slider-two-values/range-min-max-slider/index.ts [hunk 15]
```diff
@@ -1,46 +1,53 @@
...
+import { staticStyles as staticSharedStyles } from '../../../foundation/component-tokens/slider-legend.css.js';
+import { FormSizesType, ActionVariantType, RenderBtnProps } from '../../../globals/types.js';
+import {
+  findPercentage,
+  generateRangeBar,
+  findNearestValue,
+  setOnclickValue,
+} from '../../../utils/range-slider-utils.js';
+import { BlrButtonIconRenderFunction } from '../../button-icon/renderFunction.js';
+import { LitElementCustom } from '../../../utils/lit/element.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/slider-two-values/range-min-max-slider/index.ts:18]: Please integrate line 18 and 19

File: packages/ui-library/src/components/textarea/index.ts [hunk 16]
```diff
@@ -21,8 +22,9 @@ import {
   createBlrFocusEvent,
   createBlrSelectEvent,
   createBlrTextValueChangeEvent,
-} from '../../globals/events';
-import { LitElementCustom } from '../../utils/lit-element-custom';
+} from '../../globals/events.js';
+import { LitElementCustom } from '../../utils/lit/element.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/textarea/index.ts:26]: Please integrate line 26 and 27

File: packages/ui-library/src/components/toggle-switch/index.ts [hunk 17]
```diff
@@ -1,26 +1,28 @@
...
   BlrBlurEvent,
   BlrCheckedChangeEvent,
   BlrFocusEvent,
   createBlrBlurEvent,
   createBlrCheckedChangeEvent,
   createBlrFocusEvent,
-} from '../../globals/events';
-import { LitElementCustom } from '../../utils/lit-element-custom';
+} from '../../globals/events.js';
+import { LitElementCustom } from '../../utils/lit/element.js';
```
Comment (angsherpa456) [inline packages/ui-library/src/components/toggle-switch/index.ts:24]: Please integrate line 24 and 25

**Review by angsherpa456 — COMMENTED** [review]
Overall looks good. Just curious regarding the step that needs to be taken to resolve the current slow loading issue.

**Comment by faselbaum** [comment]
> Overall looks good. Just curious regarding the step that needs to be taken to resolve the current slow loading issue. Just a hint: The lit-labs/ssr states that "@lit-labs/ssr is pre-release software, not quite ready for public consumption", there might be potential breaking changes required later on if it released the stable version.

From my perspective it's still worth adopting the `@lit-labs/ssr` package now as it is the only way we can actually verify ssr capabilities with lit. SSR is still a common requirement for many projects. @thrbnhrtmnn, @ChristianHoffmannS2 and me had this discussion while you were on ADAC and we agreed to provide a test bed for further research and development. Imho not investigating ssr and later finding out aren't ssr compatible at all would be far worse than adapting to breaking changes.

Regarding the slow speed: Let's tackle this in a [follow up issue](https://github.com/deven-org/boiler/issues/1122). This PR is mainly aimed at providing a development foundation.

**Comment by thrbnhrtmnn** [comment]
> @angsherpa456
> 
> > Overall looks good. Just curious regarding the step that needs to be taken to resolve the current slow loading issue. Just a hint: The lit-labs/ssr states that "@lit-labs/ssr is pre-release software, not quite ready for public consumption", there might be potential breaking changes required later on if it released the stable version.
> 
> From my perspective it's still worth adopting the `@lit-labs/ssr` package now as it is the only way we can actually verify ssr capabilities with lit. SSR is still a common requirement for many projects. @thrbnhrtmnn, @ChristianHoffmannS2 and me had this discussion while you were on ADAC and we agreed to provide a test bed for further research and development. Imho not investigating ssr and later finding out that our components aren't ssr compatible at all would be far worse than adapting to breaking changes.
> 
> Regarding the slow speed: Let's tackle this in a [follow up issue](https://github.com/deven-org/boiler/issues/1122). This PR is mainly aimed at providing a development foundation.

Thanks @faselbaum for the follow up task, I agree with everything you wrote. We should adopt the ssr package now and potentially adopt to breaking changes in the future rather than not verifying the ssr capabilities at all.

File: packages/ui-library/src/components/select/index.ts [hunk 18]
```diff
@@ -82,6 +82,10 @@ export class BlrSelect extends LitElementCustom {
     }
   };
 
+  protected firstUpdated(_changedProperties: PropertyValueMap<any> | Map<PropertyKey, unknown>): void {
+    this.handleSlotChange();
```
Comment (angsherpa456) [inline packages/ui-library/src/components/select/index.ts:86]: Now it gets called twice, right? Once from the <slot @slotchange={..}> and once from firstUpdated.

Comment (faselbaum) [reply packages/ui-library/src/components/select/index.ts:86]: Correct. It's only a hotfix until then

**Pull request merged**